From e43f8fe284aad21f3276abec5b1ec6483af00392 Mon Sep 17 00:00:00 2001
From: Anssi Hannula <anssi.hannula@iki.fi>
Date: Sun, 17 Apr 2022 04:37:48 +0000
Subject: [PATCH] HACK: ALSA: Assign internal PCM chmap/ELD/IEC958 kctls to
 device 0

On SoC sound devices utilizing codec2codec DAI links with a HDMI codec
the kctls for chmap, ELD, IEC958 are currently created using the
internal PCM device numbers. This causes userspace to not see the
actual channel mapping.

Affected devices include LibreTech LePotato and Wetek Play 2.

The proper fix would be not create these kctls for internal PCMs and
instead create them for the real userspace-visible PCMs, somehow
forwarding the controls between the HDMI codec and the real PCM.

As a workaround, simply use device=0 for all channel map controls and
SoC HDMI codec controls for internal PCM devices.

Signed-off-by: Anssi Hannula <anssi.hannula@iki.fi>
---
 sound/core/pcm_lib.c          | 5 ++++-
 sound/soc/codecs/hdmi-codec.c | 4 +++-
 2 files changed, 7 insertions(+), 2 deletions(-)

diff --git a/sound/core/pcm_lib.c b/sound/core/pcm_lib.c
index 45afef73275f..112f89386f60 100644
--- a/sound/core/pcm_lib.c
+++ b/sound/core/pcm_lib.c
@@ -2457,7 +2457,10 @@ int snd_pcm_add_chmap_ctls(struct snd_pcm *pcm, int stream,
 		knew.name = "Playback Channel Map";
 	else
 		knew.name = "Capture Channel Map";
-	knew.device = pcm->device;
+	if (pcm->internal && pcm->device)
+		dev_info(pcm->card->dev, "workaround active: internal PCM chmap controls mapped to device 0\n");
+	else
+		knew.device = pcm->device;
 	knew.count = pcm->streams[stream].substream_count;
 	knew.private_value = private_value;
 	info->kctl = snd_ctl_new1(&knew, info);
diff --git a/sound/soc/codecs/hdmi-codec.c b/sound/soc/codecs/hdmi-codec.c
index 403d4c6a49a8..7a1ccedd730e 100644
--- a/sound/soc/codecs/hdmi-codec.c
+++ b/sound/soc/codecs/hdmi-codec.c
@@ -630,7 +630,6 @@ static int hdmi_codec_pcm_new(struct snd_soc_pcm_runtime *rtd,
 		.name	= "ELD",
 		.info	= hdmi_eld_ctl_info,
 		.get	= hdmi_eld_ctl_get,
-		.device	= rtd->pcm->device,
 	};
 	int ret;
 
@@ -648,6 +647,9 @@ static int hdmi_codec_pcm_new(struct snd_soc_pcm_runtime *rtd,
 	hcp->chmap_info->chmap = hdmi_codec_stereo_chmaps;
 	hcp->chmap_idx = HDMI_CODEC_CHMAP_IDX_UNKNOWN;
 
+	if (!rtd->pcm->internal)
+		hdmi_eld_ctl.device = rtd->pcm->device;
+
 	/* add ELD ctl with the device number corresponding to the PCM stream */
 	kctl = snd_ctl_new1(&hdmi_eld_ctl, dai->component);
 	if (!kctl)
